### ISSUE TITLE
Refactoring the AIRDock class

### ISSUE THREAD

**Issue by GimmickNG** [GimmickNG/AIRDock#3]
It's been brought to my attention that AIRDock.as ([master](../blob/master/src/airdock/AIRDock.as) | [experimental](../blob/experimental/src/airdock/AIRDock.as)) is in need of breaking down into smaller class - as a ~2000 line class it seems to be doing too many things at once.

What will follow this issue is a series of comments indicating the relationship between various functions in AIRDock.as, so as to identify regions of separation and possible cross-cutting concerns.

**Comment by GimmickNG** [comment]
The AIRDock class consists of a bunch of different functions built in:

1. Handling containers - creation, addition, removal, docking
2. Handling panels - creation, adding to containers, removing from containers
3. Docking panels and containers - starting a dock operation, continuing the dock operation, stopping the dock operation and deciding where and how to merge containers at the end (i.e. actions taken when valid or invalid dock targets)
4. Resizing containers
5. Displaying dock helpers
6. Creating panel lists for containers
7. Creating windows for containers and managing them
8. Managing the docking policy

The question now is - how to break it down into smaller classes? Overlapping functions (those dependent on each other) are:

1. Container docking
2. Docking panels and containers
3. Displaying dock helpers
4. Managing the docking policy

All the above 4 require, or are triggered by, NativeDragEvent handlers. At the same time, these require references to containers which are created by the AIRDock instance.

**Comment by GimmickNG** [comment]
Re: displaying and hiding the dock helper requiring NativeDragEvent handlers

The dock helper need not necessarily require a NativeDragEvent handler to be displayed; as a preliminary separation, the DockHelper class could be displayed only when a DOCK_REQUEST or some other event is fired by AIRDock, instead of directly calling it via the NDE handler. Something like this:

```
NativeDragEvent                         Dispatches a                     Dock handles
when the user              ---->        DOCK_REQUEST        ---->        appear on the
drags the container                     event to the dock                container (via
over another container                  handle manager                   event params)
```

**Comment by GimmickNG** [comment]
However, the native drag event handlers are registered to the containers, rather than to the AIRDock instance. This is rather tightly coupled with respect to the container because another method of docking (assume that a normal MouseEvent, instead of a NativeDragEvent, were used) would not need those listeners. What would be an extensible way of registering listeners to only those targets needed, and allow for various docking strategies at the same time?

That is, suppose NativeDragEvent gets removed (unlikely, but this is a thought exercise) in future versions of AS3, or the library gets ported to another language which doesn't support it, but supports something similar which uses slightly different methods of docking. How would this library reconcile that? It can't add all kinds of event listeners and forward them; that's too inefficient.

An alternative is to have separate classes which would register event listeners to the AIRDock instance, but the AIRDock class would transparently register those same event listeners to all the containers, windows, and other objects it manages, depending on the type of event listener that's being registered (e.g a NativeWindowBoundsEvent can only be registered for NativeWindow objects).

In short, the situation can be summarized as:

```
Docking class REGISTERS listeners ON AIRDock;
Event occurs AND AIRDock SENDS message TO docking class
```

The hard part's `REGISTERS listeners` at the moment.

**Comment by GimmickNG** [comment]
Re: Other avenues of separation

Other places where the class could be split:

- [ ] Separating the resizer from the rest of the AIRDock class. This is already done to a degree, but the logic of the resizer (show / hide, etc.) can be moved into a separate class.
- [ ] Moving utility functions into another class
- [ ] Reducing direct responsibility by shrinking feature set - for example, a panel's window's title is changed automatically whenever the panel is renamed, due to a CHANGE event handler. This could be removed and the title could be changed whenever the window's visibility is toggled instead. (Such an example is a bit extreme because it would end up contributing very little, but the principle is what should matter)

**Comment by GimmickNG** [comment]
[Here is an internal dependency graph](https://bubbl.us/NDY5Njg2MS85MDEyMjEwLzNiYmQwODQ1NmRmZDIzODIzYmE1ZTE5ZjAzMjFlYTY2-X) for the AIRDock class - bold indicates an event listener function. How should the functions be separated out? As can be seen, it's quite connected with each other:

---

![depgraph](https://user-images.githubusercontent.com/9390648/39363763-0a78c756-4a49-11e8-8acd-11b97d49b201.png)

**Comment by GimmickNG** [comment]
Since version 0.4, strategy classes have been introduced to manage different functions of the `AIRDock` (`IBasicDocker` and `ICustomizableDocker` implementation) class.
